api/v1alpha4/clusterclass_webhook.go

```diff
@@ -132,6 +161,107 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
 			)
 		}
 	}
+	return allErrs
+}
+
+func (in ClusterClass) validateRefTemplates() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if !strings.HasSuffix(strings.ToLower(in.Spec.Infrastructure.Ref.Kind), "template") {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec", "infrastructure", "ref", "kind"),
+				in.Spec.Infrastructure.Ref.Kind,
+				"must be a template",
+			),
+		)
+	}
+	if !strings.HasSuffix(strings.ToLower(in.Spec.ControlPlane.Ref.Kind), "template") {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec", "controlplane", "ref", "kind"),
+				in.Spec.ControlPlane.Ref.Kind,
+				"must be a template",
+			),
+		)
+	}
+	for _, class := range in.Spec.Workers.MachineDeployments {
+		if !strings.HasSuffix(strings.ToLower(class.Template.Bootstrap.Ref.Kind), "template") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments", "template", "bootstrap", "ref", "kind"),
+					class.Template.Bootstrap.Ref.Kind,
+					"must be a template",
+				),
+			)
+		}
+		if !strings.HasSuffix(strings.ToLower(class.Template.Infrastructure.Ref.Kind), "template") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments", "template", "infrastructure", "ref", "kind"),
+					class.Template.Infrastructure.Ref.Kind,
+					"must be a template",
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func (in ClusterClass) validateRefAPIVersions() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if in.Spec.Infrastructure.Ref.APIVersion != "" {
+		if _, err := schema.ParseGroupVersion(in.Spec.Infrastructure.Ref.APIVersion); err != nil {
```

Comment:
Should we check if the parsed value is not an empty GroupVersion?

Reply:
Since we are already checking for an empty APIVersion in line 214 I dont think `ParseGroupVersion` will return an empty GroupVersion unless there is an error (which we are already checking for).
And since `APIVersion` is technically [optional](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5422-L5424) in an `ObjectReference` we need the check in line 214.

Reply:
Looking at the conversation here I think we expect that at least `namespace`, `name`, `kind` and `apiVersion` are set on the object reference and we are supposed to throw an error if at least of them is not set, correct?

Reply:
yes! we need all of them (although the user can skip to set the namespace, which is defaulted if missing)